TinyPG/Compiler/TerminalSymbol.cs

```diff
@@ -27,7 +27,25 @@ public TerminalSymbol(string name)
         public TerminalSymbol(string name, string pattern)
         {
             Name = name;
-            Expression = new Regex(pattern, RegexOptions.Compiled);
+
+            // The pattern is a lot faster if the Regex starts on a caret,
+            // as per PR #4:  Since the scanner throws away any matches
+            // that don't start at index zero anyway, there is no logical
+            // difference between having the caret and not having it - just
+            // a speed difference.  BUT, be careful to only insert that
+            // caret if the value in the grammar file was actually a string
+            // literal rather than a bit of code or a variable identifier
+            // containing the pattern:
+            int insertIndex = 0;
+            string patternWithCaret = pattern;
+            if (pattern.StartsWith("@\"")) // string literal starting with @"
+                insertIndex = 2;
+            else if (pattern.StartsWith("\"")) // string literal starting with "
+                insertIndex = 1;
+            if (pattern[insertIndex] != '^')
+                patternWithCaret = pattern.Insert(insertIndex, "^");
```

Comment:
I think here we actually want to wrap the entire existing expression into a group `^(existing_regex)` because if the regex is something like `true|false` then adding a caret at the start will only enforce that `true` is at the beginning of the line, but false could be anywyere. We'll actually probably want to use a non-capturing group in case they are using the group indexes for matching, so like `^(?:existing_regex)`

Reply:
Okay that's a good idea.